controller/src/main/java/io/pravega/controller/server/rest/resources/StreamMetadataResourceImpl.java

```diff
@@ -563,11 +568,55 @@ public void listStreams(final String scopeName, final String showInternalStreams
             LoggerHelpers.traceLeave(log, "listStreams", traceId);
             return;
         }
-        boolean showOnlyInternalStreams = showInternalStreams != null && showInternalStreams.equals("true");
-        controllerService.listStreamsInScope(scopeName, requestId)
-                .thenApply(streamsList -> {
-                    StreamsList streams = new StreamsList();
-                    streamsList.forEach((stream, config) -> {
+        boolean showOnlyInternalStreams = filterType != null && filterType.equals("showInternalStreams");
+        boolean showStreamsWithTag = filterType != null && filterType.equals("tag");
+        String tag = null;
+        if (showStreamsWithTag && filterValue != null) {
+            tag = filterValue;
+        }
+        if (tag != null) {
+            List<Stream> streams = new ArrayList<>();
+            String finalTag = tag;
+            localController.listStreamsForTag(scopeName, tag).collectRemaining(streams::add).thenCompose(v -> {
+                List<CompletableFuture<ImmutablePair<Stream, StreamConfiguration>>> streamConfigFutureList = streams.stream().filter(stream -> {
+                    boolean isAuthorized = false;
+                    try {
+                        isAuthorized = restAuthHelper.isAuthorized(authHeader, authorizationResource.ofStreamInScope(scopeName, stream.getStreamName()),
+                                principal, READ);
+                    } catch (AuthException e) {
+                        log.warn(requestId, "List Streams with tag {} for scope {} failed due to authentication failure.",
+                                finalTag, scopeName);
+                        // Ignore. This exception occurs under abnormal circumstances and not to determine
+                        // whether the user is authorized. In case it does occur, we assume that the user
+                        // is unauthorized.
+                    }
+                    return isAuthorized;
+                }).map(stream -> localController.getStreamConfiguration(scopeName, stream.getStreamName())
+                        .thenApply(config -> new ImmutablePair<>(stream, config)))
+                        .collect(Collectors.toList());
+                return Futures.allOfWithResults(streamConfigFutureList);
+            }).thenApply(streamConfigPairs -> {
+                StreamsList responseStreams = new StreamsList();
+                responseStreams.setStreams(new ArrayList<>());
```

Comment:
The `setStreams `is to prevent an NPE when there are no streams with the given tag. 
If a new scope is created, calling `listStreams` on the scope results in an NPE for the list streams without the tag option.
This is because `StreamsList` class initializes the list inside it as null.